abdiff/core/build_ab_images.py

```diff
@@ -1 +1,67 @@
-"""abdiff.core.build_ab_images"""
+import logging
+import os
+import shutil
+
+import docker
+from pygit2 import clone_repository
+from pygit2.enums import ResetMode
+
+from abdiff.core.utils import update_or_create_job_json
+
+logger = logging.getLogger(__name__)
+
+
+def build_ab_images(
+    job_directory: str,
+    docker_client: docker.client.DockerClient,
```

Comment:
This is a shift from what we've previously discussed but I figure the same docker client would be used at here and in the subsequent `run_ab_transforms` so the Docker client could be configured once at the CLI-level and passed to each. But let me know if this is a bad idea!

Reply:
This is a question that's come up whenever I write code that involves `boto3` clients, so thank you for bringing this up and looking forward to discussing as a team. I came across [this article](https://www.digitalocean.com/community/questions/should-i-create-a-new-instance-of-boto3-client-for-each-file-upload-request-or-use-a-shared-instance), which recommends **reusing** client instances as "each new client instance consumes resources (e.g., memory, file descriptors for connections)" and "creating a new client for every request can add significant overhead". I also tried asking this question to ChatGPT, and this was the response:
> Creating a new DockerClient instance involves establishing a new connection to the Docker daemon, which can be resource-intensive. By reusing a single instance, you minimize the overhead associated with repeatedly opening and closing connections.

I am leaning towards reusing clients (as that is what I'm also seeing in various StackOverflow threads / articles), but interested in what @ghukill 's thinks as well!

Reply:
Interesting, yeah, glad we're discussing!

My knee-jerk is that I'm a little wary having the CLI be responsible for providing a docker client, which makes this function somewhat dependent on the CLI context.  

That said... becuase you have it as a typed argument, it's _very_ clear what it needs to run.

What about a bit of a hybrid:
1. keep the argument, but make it optional/positional with a default of `None`
2. if `None` within the method call, then instantiate then via `docker_client = docker.from_env()`

To me that advantages are:
- from the CLI context, you can initialize it and pass to any functions that may need it, knowing you've prevented them from creating their own
- this function still works simply without the the CLI context

Reply:
That works for me, best of both worlds!